api/node/src/types/model.rs

```diff
@@ -1,271 +1,129 @@
+    Ok(External::new(Default::default()))
+}
 
-        let notify = JsSlintModelNotify { model: Rc::downgrade(&js_model) };
+#[napi]
+pub fn js_model_notify_row_data_changed(notify: External<SharedModelNotify>, row: u32) {
+    notify.row_changed(row as usize);
+}
 
-        js_model.model.get::<Object>()?.set("notify", notify)?;
+#[napi]
+pub fn js_model_notify_row_added(notify: External<SharedModelNotify>, row: u32, count: u32) {
+    notify.row_added(row as usize, count as usize);
+}
 
-        Ok(js_model)
-    }
+#[napi]
+pub fn js_model_notify_row_removed(notify: External<SharedModelNotify>, row: u32, count: u32) {
+    notify.row_removed(row as usize, count as usize);
+}
 
-    pub fn model(&self) -> &RefCountedReference {
-        &self.model
-    }
+#[napi]
+pub fn js_model_notify_reset(notify: External<SharedModelNotify>) {
+    notify.reset();
 }
 
 impl Model for JsModel {
     type Data = slint_interpreter::Value;
 
     fn row_count(&self) -> usize {
-        let model: Object = self.model.get().unwrap();
+        let model: Object = self.js_impl.get().unwrap();
         model
             .get::<&str, JsFunction>("rowCount")
             .ok()
             .and_then(|callback| {
-                callback.and_then(|callback| callback.call::<JsUnknown>(Some(&model), &[]).ok())
+                callback.and_then(|callback| callback.call_without_args(Some(&model)).ok())
             })
             .and_then(|res| res.coerce_to_number().ok())
             .map(|num| num.get_uint32().ok().map_or(0, |count| count as usize))
             .unwrap_or_default()
     }
 
     fn row_data(&self, row: usize) -> Option<Self::Data> {
-        let model: Object = self.model.get().unwrap();
-        model
+        let model: Object = self.js_impl.get().unwrap();
+        let row_data_fn = model
             .get::<&str, JsFunction>("rowData")
-            .ok()
-            .and_then(|callback| {
-                callback.and_then(|callback| {
-                    callback
-                        .call::<JsNumber>(
-                            Some(&model),
-                            &[self.env.create_double(row as f64).unwrap()],
-                        )
-                        .ok()
-                })
-            })
-            .and_then(|res| {
-                if res.get_type().unwrap() == ValueType::Undefined {
-                    None
-                } else {
-                    to_value(&self.env, res, &self.data_type).ok()
-                }
-            })
+            .expect("Node.js: JavaScript Model<T> implementation is missing rowData property")
+            .expect("Node.js: Model<T> implementation's rowData property is not a function");
+        let row_data = row_data_fn
+            .call::<JsNumber>(Some(&model), &[self.env.create_double(row as f64).unwrap()])
+            .expect("Node.js: JavaScript Model<T>'s rowData function threw an exception");
```

Comment:
This is annoying. I'm not sure it is realy wise to panic here. Bu i don't know what to do. Should we return None and somehow log the error or rethrow the exception?

Reply:
That's a good idea for an alternative, indeed. (Return None and log). Previously we just unwrapped.